### PR TITLE
Improve documentation and code quality

### PR DESCRIPTION
**Pull request by froozen**
That was quite a lot of hacky code.
Please check the comments make sure I described everything correctly.

### CODE REVIEW

**Comment by froozen** [comment]
Also, I noticed `isClient` isn't used outside of `initConnection`. Afterwards the spec says it's irrelevant whether the client is the client or server side. Can I remove it from `Connection` or should I leave it?

File: src/Demo/Server.hs [hunk 1]
```diff
@@ -17,7 +17,7 @@ import           Network                     hiding (connectTo)
 main = do
   state <- createState (PortNumber 9878)
   flip (runStateT . runRicochet) state $ do
-    versions %= (M.insert 1 (\_ -> return ()))
+    versions %= M.insert 1 closeConnection
```
Comment (sternenseemann) [inline src/Demo/Server.hs]: The intention here was not to close the connection but to do nothing and execute the rest of the `do` block instead.

File: src/Network/Ricochet/Connection.hs [hunk 2]
```diff
@@ -24,53 +25,90 @@ import           Network.Socks5           (socksConnectTo)
...
-        let chosen = foldl1 max (fmap fst handlers)
-        liftIO . putStrLn $ "We can choose between " <> show (length handlers) <> " versions!"
-        liftIO . putStrLn $ "We have chosen " <> show chosen <> "."
-        liftIO . B.hPutStr handle $ B.singleton chosen
-        connections . traverse . filtered (== con) . cInputBuffer %= (<> rest)
-        (fromJust $ lookup chosen handlers) con
-      Nothing -> liftIO $ putStrLn "Remote side sent invalid version negotiation."
-  return con
+  -- Create the actual connection structure
+  let channels = [MkChannel 0 $ MkChannelType "im.ricochet.control-channel"]
```
Comment (sternenseemann) [inline src/Network/Ricochet/Connection.hs]: Maybe a wrapper function similar to `makePacket` would be useful here?

File: src/Network/Ricochet/Connection.hs [hunk 3]
```diff
@@ -24,53 +25,90 @@ import           Network.Socks5           (socksConnectTo)
...
+--   handler corresponding to the one the peer picked
+offerVersions :: Connection -> Ricochet ()
+offerVersions connection = do
+  availableVersions <- use versions
+  -- Send the available versions
+  liftIO . B.hPutStr (connection ^. cHandle) $ dumpIntroduction availableVersions
+  response <- liftIO $ B.hGet (connection ^. cHandle) 1
+  let choice = head $ B.unpack response
+  -- If the choice is valid
+  if choice `M.member` availableVersions
```
Comment (sternenseemann) [inline src/Network/Ricochet/Connection.hs:83]: We should explicitly check for the `0xFF` here since that version could be part of the `Map` if somebody used the library wrong.

File: src/Network/Ricochet/Connection.hs [hunk 4]
```diff
@@ -24,53 +25,90 @@ import           Network.Socks5           (socksConnectTo)
...
+      liftIO . B.hPutStr (connection ^. cHandle) $ B.singleton 0xFF
+      closeConnection connection
+    -- The versions match
+    Just (handlers, rest) -> do
+      -- Always choose the latest version
+      let chosen = maximum (fmap fst handlers)
+      liftIO . B.hPutStr (connection ^. cHandle) $ B.singleton chosen
+      -- Append the rest of the response to the inputBuffer
+      connections . traverse . filtered (== connection) . cInputBuffer %= (<> rest)
+      -- Start the chosen hanler
```
Comment (sternenseemann) [inline src/Network/Ricochet/Connection.hs]: Typo here: `handler`, not `hanler`.

**Comment by sternenseemann** [comment]
If you are right about that then `isClient` could be removed from `Connection`.

Nice work! You were able to clean up the code and add meaningful documentation.

**Comment by froozen** [comment]
> We should explicitly check for the 0xFF here since that version could be part of the Map if somebody used the library wrong.

This should be done in the function for starting the Ricochet monad instead of inside the implementation.

**Comment by froozen** [comment]
Any more comments? This should be ready to be merged otherwise.

**Comment by sternenseemann** [comment]
> This should be done in the function for starting the Ricochet monad instead of inside the implementation.

What do you mean by that?

**Comment by froozen** [comment]
Well, to run haskell-ricochet, the user will have to execute a function that takes the version Map. This function should make sure that looking up `0xFF` returns `Nothing`.

**Comment by sternenseemann** [comment]
Right.

**Comment by photm5** [comment]
We do need to know whether we’re the client side, as the client is required to prove his control over a hidden service.  In addition, a user of the library might choose to implement a custom protocol version that uses the information somewhere.

File: src/Network/Ricochet/Connection.hs [hunk 5]
```diff
@@ -24,53 +25,88 @@ import           Network.Socks5           (socksConnectTo)
...
-connectTo :: String -> PortID -> Ricochet Connection
+-- | Connects to a peer through the Tor network
+connectTo :: String              -- ^ Tor hidden service identifier to connect to
+          -> PortID              -- ^ Port to connect to
+          -> Ricochet Connection -- ^ Returns the established connection
 connectTo domain port = do
   torPort <- use socksPort
   handle <- liftIO $ socksConnectTo "localhost" torPort domain port
-  initConnection handle True
+  initiateConnection handle False
```
Comment (photm5) [inline src/Network/Ricochet/Connection.hs:48]: Why did you change `True` to `False`?

File: src/Network/Ricochet/Connection.hs [hunk 6]
```diff
@@ -24,53 +25,88 @@ import           Network.Socks5           (socksConnectTo)
...
 
-initConnection :: Handle -> Bool -> Ricochet Connection
-initConnection handle isClientSide = do
+-- | Initiate a newly made connection
+initiateConnection :: Handle              -- ^ Handle corresponding to the connection to the peer
+                   -> Bool                -- ^ Wether this side accepted the connection
```
Comment (photm5) [inline src/Network/Ricochet/Connection.hs:52]: This is exactly the opposite of what `isClientSide` says.  It should be “Whether this side opened the connection.”

File: src/Network/Ricochet/Protocol/Lowest.hs [hunk 7]
```diff
@@ -38,14 +38,24 @@ packet = do
...
+                  -> [Packet]   -- ^ Returns a list of packets containing the ByteString
+splitIntoPackets chan bs = let (ps, bs') = splitIntoPackets' chan ([], bs)
+                           in  ps <> [makePacket chan bs']
+
+-- | Iteratively turn a ByteString into a series of maximal sized packets and
+--   return the rest
+splitIntoPackets' :: Word16                 -- ^ ID of the channel the packets should be sent on
+                  -> ([Packet], ByteString) -- ^ A tuple of the packets created so far and the
+                                            --   ByteString to conusme
+                  -> ([Packet], ByteString) -- ^ Rerturns a tuple of the packets created and a
```
Comment (photm5) [inline src/Network/Ricochet/Protocol/Lowest.hs:53]: Should be “Returns.”

File: src/Network/Ricochet/Protocol/Lowest.hs [hunk 8]
```diff
@@ -38,14 +38,24 @@ packet = do
...
+                  -> ByteString -- ^ The ByteString to be split
+                  -> [Packet]   -- ^ Returns a list of packets containing the ByteString
+splitIntoPackets chan bs = let (ps, bs') = splitIntoPackets' chan ([], bs)
+                           in  ps <> [makePacket chan bs']
+
+-- | Iteratively turn a ByteString into a series of maximal sized packets and
+--   return the rest
+splitIntoPackets' :: Word16                 -- ^ ID of the channel the packets should be sent on
+                  -> ([Packet], ByteString) -- ^ A tuple of the packets created so far and the
+                                            --   ByteString to conusme
```
Comment (photm5) [inline src/Network/Ricochet/Protocol/Lowest.hs:52]: Should be “consume.”

File: src/Network/Ricochet/Protocol/Lowest.hs [hunk 9]
```diff
@@ -38,14 +38,24 @@ packet = do
...
+                  -> ([Packet], ByteString) -- ^ Rerturns a tuple of the packets created and a
+                                            --   ByteString shorter than the maximal packet length
+splitIntoPackets' chan (ps, bs) =
+    case (toInteger . B.length $ bs) `compare` toInteger maxPackLen of
       EQ -> (ps <> [makePacket chan bs], B.empty)
       LT -> (ps, bs)
-      GT -> splitInPackets' chan (ps <> [makePacket chan (B.take packLen bs)], B.drop packLen bs)
-  where packLen = fromIntegral (maxBound :: Word16) - 4
+      -- Call the function again if there is some ByteString left
```
Comment (shak-mar) [inline src/Network/Ricochet/Protocol/Lowest.hs]: Should be “if there is *enough* ByteString left.”

File: src/Network/Ricochet/Types.hs [hunk 10]
```diff
@@ -15,9 +16,11 @@ data Packet = MkPacket
   , _pPacketData :: ByteString
   } deriving (Show)
 
--- | Function creating a packet with appropiate size from a Channel and a ByteString
-makePacket :: Word16 -> ByteString -> Packet
-makePacket chan bs = MkPacket (4 + (fromIntegral $ B.length bs)) chan bs
+-- | Creates a packet with appropiate size from a Channel and a ByteString
```
Comment (photm5) [inline src/Network/Ricochet/Types.hs:19]: Should be “appropriate.”

File: src/Network/Ricochet/Types.hs [hunk 11]
```diff
@@ -27,10 +30,15 @@ makePacket chan bs = MkPacket (4 + (fromIntegral $ B.length bs)) chan bs
...
   , _cInputBuffer :: ByteString
   }
 
+-- | Creates an initial 'Connection' from a Handle
+makeConnection :: Handle -> Connection
+makeConnection handle = -- Start out with only a Control Channel
+  let channels = [MkChannel 0 $ MkChannelType "im.ricochet.control-channel"]
+  in  MkConnection handle channels B.empty
+
 instance Eq Connection where
```
Comment (photm5) [inline src/Network/Ricochet/Types.hs:42]: We should add a note explaining that while this `Eq` instance obeys the laws, it does not check for equality on every record member.

File: src/Network/Ricochet/Connection.hs [hunk 12]
```diff
@@ -24,53 +25,88 @@ import           Network.Socks5           (socksConnectTo)
 import           System.IO                (BufferMode (..), Handle (),
                                            hSetBuffering)
 
+-- | Creates a new RicochetState listening on the supplied port
 createState :: PortID -> IO RicochetState
 createState port = do
   sock <- listenOn port
   return $ MkRicochetState sock [] [] (PortNumber 9050) M.empty
 
+-- | Waits until a new peer connects to initiae a connection
```
Comment (photm5) [inline src/Network/Ricochet/Connection.hs:34]: Should be “initiate.”

File: src/Network/Ricochet/Monad.hs [hunk 13]
```diff
@@ -45,19 +47,23 @@ data RicochetState = MkRicochetState
...
 peekPacket :: Connection -> Ricochet (Maybe Packet)
 peekPacket con = do
-  readBytes <- liftIO $ B.hGetNonBlocking (con ^. cHandle) max
+  readBytes <- liftIO $ B.hGetNonBlocking (con ^. cHandle) maxPacketSize
+  -- Append the read bytes to the input buffer
   inputBuffer <- con' . cInputBuffer <%= (<> readBytes)
+  -- Try parsing a full packet and return it on success
   case parsePacket inputBuffer of
     Success packet bs -> do
+      -- Removed the parsed portion from the inputBuffer
```
Comment (photm5) [inline src/Network/Ricochet/Monad.hs:59]: Should be imperative instead of past tense: “Remove”

File: src/Network/Ricochet/Monad.hs [hunk 14]
```diff
@@ -74,5 +80,14 @@ sendPacket :: Connection -> Packet -> Ricochet ()
 sendPacket con pkt = liftIO . B.hPutStr (con ^. cHandle) $ dumpPacket pkt
 
 -- | Packs a ByteString into Packets and sends it through the given Connection
-sendByteString :: Connection -> Word16 -> ByteString -> Ricochet ()
-sendByteString con chan bs = mapM_ (sendPacket con) $ splitInPackets chan bs
+sendByteString :: Connection -- ^ The Connection through which to send the ByteString
+               -> Word16     -- ^ The ID of the channel the ByteString should be send on
```
Comment (photm5) [inline src/Network/Ricochet/Monad.hs:84]: Should be “sent.”

File: src/Network/Ricochet/Monad.hs [hunk 15]
```diff
@@ -74,5 +80,14 @@ sendPacket :: Connection -> Packet -> Ricochet ()
...
+               -> ByteString -- ^ The ByteString to be sent
+               -> Ricochet ()
+sendByteString con chan bs = mapM_ (sendPacket con) $ splitIntoPackets chan bs
+
+-- | Close a connection and remove it from the gg
```
Comment (photm5) [inline src/Network/Ricochet/Monad.hs:89]: What’s the “gg?”

File: src/Network/Ricochet/Monad.hs (hunk 13 above)
Comment (sternenseemann) [inline src/Network/Ricochet/Monad.hs:59]: Should be simple "present instead" of "imperative" :new_moon_with_face:

Comment (photm5) [inline src/Network/Ricochet/Monad.hs:59]: What?

File: src/Network/Ricochet/Connection.hs (hunk 5 above)
Comment (sternenseemann) [inline src/Network/Ricochet/Connection.hs:48]: Fun fact: Everything seems still to work after merge.

Comment (froozen) [inline src/Network/Ricochet/Connection.hs:48]: I guess that just goes to show that a magic boolean isn't the way to go. I'll replace it with a custom data type.

File: src/Network/Ricochet/Types.hs (hunk 11 above)
Comment (froozen) [inline src/Network/Ricochet/Types.hs:42]: It's explained in the documentation of `Connection`. Adding the same to the `Eq` instance is of little use as it won't be included in the haddock documentation.

File: src/Network/Ricochet/Monad.hs (hunk 13 above)
Comment (froozen) [inline src/Network/Ricochet/Monad.hs:59]: I think he means "Removes" instead of "Remove"

File: src/Network/Ricochet/Types.hs (hunk 11 above)
Comment (photm5) [inline src/Network/Ricochet/Types.hs:42]: First startpage result for “haddock instance comments”: http://trac.haskell.org/haddock/ticket/29. I guess that fix will have made it into the versions we’re using. I remember seeing it in some haddock documentation already.

File: src/Network/Ricochet/Monad.hs (hunk 13 above)
Comment (photm5) [inline src/Network/Ricochet/Monad.hs:59]: That wouldn’t fit with all the other comments though.

File: src/Network/Ricochet/Connection.hs (hunk 6 above)
Comment (sternenseemann) [inline src/Network/Ricochet/Connection.hs:52]: indeed

File: src/Network/Ricochet/Connection.hs (hunk 5 above)
Comment (sternenseemann) [inline src/Network/Ricochet/Connection.hs:48]: @froozen :+1:
